File: lane_detection/src/image_processor/image_processor_factory/image_processor_factory.py
import logging

from image_processor.impl.image_processor_impl import ImageProcessorImpl
from image_processor.image_processor_factory.enums import ImageProcessorImplementation

from utils.loadConfig import loadConfig

CONFIG = loadConfig('config.json')


class ImageProcessorFactory():
    def createImageProcessor(self):
        if CONFIG['imageProcessorImplementation'] == ImageProcessorImplementation.IMG_PROC:
            imageProcessor = ImageProcessorImpl()

        else:
            raise ValueError(
                f'{__class__.__name__}: OutputImplementation - Choose from {list(ImageProcessorImplementation)}')

        logging.info(
            f'{__class__.__name__} - Created ImageProcessor: {imageProcessor.__class__.__name__}')

        return imageProcessor

File: lane_detection/src_cpp/image_processor_factory.cpp
#include "image_processor_factory.h"
#include "ImageProcessorImpl.h"

ImageProcessorImpl ImageProcessorFactory::createImageProcessor() {
    // auto config = loadConfig("config.json");
    // ImageProcessorImplementation implementation = config.imageProcessorImplementation;
    // ImageProcessorImplementation implementation = ImageProcessorImplementation::IMG_PROC;
    
    ImageProcessorImpl imageProcessor;

    // if (implementation == ImageProcessorImplementation::IMG_PROC) {
    //     ImageProcessorImpl imageProcessor;
    // } else {
    //     throw std::invalid_argument("Invalid Image Processor Implementation");
    // }

    std::cout << "ImageProcessorFactory - Created ImageProcessor: " << typeid(imageProcessor).name() << std::endl;

    return imageProcessor;
}
